# ЗАДАНИЕ 9.5. МОДУЛЬ ML-6 (HW-03)
Обучите модель линейной регрессии на найденных двумя способами трёх важных признаках и сравните полученные результаты. Загрузите полученный ноутбук (в формате IPYNB) в форму ниже.

In [33]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error
from sklearn import metrics
import numpy as np


# ПОДГОТОВКА ДАННЫХ

In [2]:
%%capture
!wget https://www.dropbox.com/s/64ol9q9ssggz6f1/data_ford_price.xlsx

In [3]:
data = pd.read_excel('data_ford_price.xlsx') 

In [4]:
data = data[['price','year', 'cylinders', 'odometer', 'lat', 'long', 'weather']]
data.dropna(inplace = True)

y = data['price']
x = data.drop(columns='price')

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=40)

In [5]:
model = LinearRegression()
model.fit(X_train, y_train)
y_predicted = model.predict(X_test) 
mae = mean_absolute_error(y_test, y_predicted)
print('MAE: %.3f' % mae)

MAE: 4682.957


# RFE

In [6]:
from sklearn.feature_selection import RFE

In [7]:
estimator = LinearRegression()
selector = RFE(estimator, n_features_to_select=3, step=1)
selector = selector.fit(X_train, y_train)
 
selector.get_feature_names_out()

array(['year', 'cylinders', 'lat'], dtype=object)

Обучим модель на выделенных признаках для RFE

In [31]:
model.fit(X_train[['year', 'cylinders', 'lat']], y_train)
y_predicted = model.predict(X_test[['year', 'cylinders', 'lat']]) 
# Рассчитываем MAE - Средняя абсолютная ошибка
print('MAE score: {:.3f}'.format(metrics.mean_absolute_error(y_test, y_predicted)))
# Рассчитываем коэффициент детерминации Коэффициент детерминации (R2)
print('R2 score: {:.3f}'.format(metrics.r2_score(y_test, y_predicted)))

MAE score: 5096.570
R2 score: 0.576


# SelectKBest

In [16]:
from sklearn.feature_selection import SelectKBest, f_regression, f_classif

In [17]:
selector = SelectKBest(f_regression, k=3)
selector.fit(X_train, y_train)
 
selector.get_feature_names_out()

array(['year', 'cylinders', 'odometer'], dtype=object)

Обучим модель на выделенных признаках для SelectKBest

In [32]:
model.fit(X_train[['year', 'cylinders', 'odometer']], y_train)
y_predicted = model.predict(X_test[['year', 'cylinders', 'odometer']]) 

# Рассчитываем MAE - Средняя абсолютная ошибка
print('MAE score: {:.3f}'.format(metrics.mean_absolute_error(y_test, y_predicted)))
# Рассчитываем коэффициент детерминации Коэффициент детерминации (R2)
print('R2 score: {:.3f}'.format(metrics.r2_score(y_test, y_predicted)))

MAE score: 4708.946
R2 score: 0.602


***Какой метод отбора признаков показал наилучший результат на тестовой выборке?***

По полученным метрикам для:
* RFE с признаками ['year', 'cylinders', 'lat']:
    * MAE score: 5096.570
    * R2 score: 0.576
```
```
* SelectKBest с признаками ['year', 'cylinders', 'odometer']:
    * MAE score: 4708.946
    * R2 score: 0.602

Можно сделать вывод, что наилучший результат на тестовой выборке показал метод отбора признаков **"SelectKBest"**, так как он дает меньшую метрику MAE (средняя абсолютная ошибка) и большую метрику R2 (коэф. детерминации).

